cirq-rigetti/cirq_rigetti/aspen_device.py

```diff
@@ -0,0 +1,514 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, cast, Optional, Union, Dict, Any
+import functools
+from math import sqrt
+import httpx
+import numpy as np
+import networkx as nx
+import cirq
+from pyquil.quantum_processor import QCSQuantumProcessor
+from qcs_api_client.models import InstructionSetArchitecture
+from qcs_api_client.operations.sync import (
+    get_instruction_set_architecture,
+)
+from cirq_rigetti._qcs_api_client_decorator import _provide_default_client
+
+
+class UnsupportedQubit(ValueError):
+    pass
+
+
+class UnsupportedRigettiQCSOperation(ValueError):
+    pass
+
+
+class UnsupportedRigettiQCSQuantumProcessor(ValueError):
+    pass
+
+
+_grid_qubit_mapping = {
+    cirq.GridQubit(0, 0): 2,
+    cirq.GridQubit(0, 1): 15,
+    cirq.GridQubit(1, 0): 1,
+    cirq.GridQubit(1, 1): 16,
+}
+_forward_line_qubit_mapping = [5, 4, 3, 2]
+_reverse_line_qubit_mapping = [1, 0, 7, 6]
+
+
+@cirq.value.value_equality
+class RigettiQCSAspenDevice(cirq.devices.Device):
```

Comment:
a follow-up PR idea: we could have a similar ASCII diagram drawing as the Google devices do.